dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs

```diff
@@ -0,0 +1,537 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using System.Text.Json.Serialization;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.AI.Embeddings;
+using Microsoft.SemanticKernel.Connectors.Memory.Weaviate.Diagnostics;
+using Microsoft.SemanticKernel.Connectors.Memory.Weaviate.Http.ApiSchema;
+using Microsoft.SemanticKernel.Connectors.Memory.Weaviate.Model;
+using Microsoft.SemanticKernel.Memory;
+
+namespace Microsoft.SemanticKernel.Connectors.Memory.Weaviate;
+
+/// <summary>
+/// An implementation of <see cref="IMemoryStore" /> for Weaviate.
+/// </summary>
+/// <remarks>The Embedding data is saved to Weaviate instance specified in the constructor.
+/// The embedding data persists between subsequent instances and has similarity search capability.
+/// </remarks>
+// ReSharper disable once ClassWithVirtualMembersNeverInherited.Global
+public class WeaviateMemoryStore : IMemoryStore, IDisposable
+{
+    // Regex to ensure Weaviate class names confirm to the naming convention
+    // https://weaviate.io/developers/weaviate/configuration/schema-configuration#class
+    private static readonly Regex s_classNameRegEx = new("[^0-9a-zA-Z]+", RegexOptions.Compiled);
+
+    private static readonly JsonSerializerOptions s_jsonSerializerOptions = new()
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+    };
+
+    private readonly HttpClient _httpClient;
+    private readonly bool _isSelfManagedHttpClient;
+    private readonly ILogger _log;
+    private bool _disposed;
+
+    /// <summary>
+    ///     Constructor for a memory store backed by Weaviate
+    /// </summary>
+    public WeaviateMemoryStore(string scheme, string host, int port, string? apiKey = null, HttpClient? httpClient = null, ILogger? logger = null)
+    {
+        Verify.ArgNotNullOrEmpty(scheme, "Scheme cannot be null or empty");
+        Verify.ArgNotNullOrEmpty(host, "Host cannot be null or empty");
+
+        this._log = logger ?? NullLogger<WeaviateMemoryStore>.Instance;
+        if (httpClient == null)
+        {
+            this._httpClient = new();
+            if (!string.IsNullOrEmpty(apiKey))
+            {
+                this._httpClient.DefaultRequestHeaders.Add("authorization", apiKey);
```

Comment:
I don't know how helpful it is to presume the environment in which this will be deployed - this could be helpful for some and unhelpful for others? Is it not the responsibility of the user to ensure they are taking the necessary precautions when configuring their deployment?